nvflare/private/fed/app/simulator/simulator_runner.py

```diff
@@ -0,0 +1,383 @@
+from nvflare.fuel.common.multi_process_executor_constants import CommunicationMetaData
+from nvflare.fuel.hci.server.authz import AuthorizationService
+from nvflare.fuel.hci.zip_utils import convert_legacy_zip, split_path, unzip_all_from_bytes, zip_directory_to_bytes
+from nvflare.fuel.sec.audit import AuditService
+from nvflare.private.defs import AppFolderConstants
+from nvflare.private.fed.app.deployer.simulator_deployer import SimulatorDeployer
+from nvflare.private.fed.client.client_status import ClientStatus
+from nvflare.private.fed.server.job_meta_validator import JobMetaValidator
+from nvflare.private.fed.simulator.simulator_app_runner import SimulatorServerAppRunner
+from nvflare.private.fed.simulator.simulator_const import SimulatorConstants
+from nvflare.private.fed.utils.fed_utils import add_logfile_handler
+from nvflare.security.security import EmptyAuthorizer
+
+
+class SimulatorRunner(FLComponent):
+    def __init__(self, args):
+        super().__init__()
+        self.args = args
+        self.ask_to_stop = False
+
+        self.simulator_root = None
+        self.services = None
+        self.deployer = SimulatorDeployer()
+        self.client_names = []
+
+    def setup(self):
+        if self.args.client_list:
+            self.client_names = self.args.client_list.strip().split(",")
+        elif self.args.clients:
+            for i in range(self.args.clients):
+                self.client_names.append("site-" + str(i + 1))
+
+        log_config_file_path = os.path.join(self.args.workspace, "startup", "log.config")
+        if not os.path.isfile(log_config_file_path):
+            log_config_file_path = os.path.join(os.path.dirname(__file__), "resource/log.config")
+        logging.config.fileConfig(fname=log_config_file_path, disable_existing_loggers=False)
+
+        # self.logger = logging.getLogger()
+        self.args.log_config = None
+        self.args.config_folder = "config"
+        self.args.job_id = SimulatorConstants.JOB_NAME
+        self.args.client_config = os.path.join(self.args.config_folder, "config_fed_client.json")
+        self.args.env = os.path.join("config", AppFolderConstants.CONFIG_ENV)
+        cwd = os.getcwd()
+        self.args.job_folder = os.path.join(cwd, self.args.job_folder)
+
+        if not os.path.exists(self.args.workspace):
+            os.makedirs(self.args.workspace)
+        os.chdir(self.args.workspace)
+        AuthorizationService.initialize(EmptyAuthorizer())
+        AuditService.initialize(audit_file_name=WorkspaceConstants.AUDIT_LOG)
+
+        self.simulator_root = os.path.join(self.args.workspace, SimulatorConstants.JOB_NAME)
+        if os.path.exists(self.simulator_root):
+            shutil.rmtree(self.simulator_root)
+
+        try:
+            data_bytes, job_name, meta = self.validate_job_data()
+
+            if not self.client_names:
+                self.client_names = self._extract_client_names_from_meta(meta)
+            if not self.client_names:
```

Comment:
Yes, if meta uses "@all", and provides no client name option, it will result in error message:
"Please provide the client names list, or the number of clients to run the simulator"